Add support for the oidc introspection endpoint

jvlcek opened the pull request:
Fixes ManageIQ/manageiq#19866

This PR adds support for the oidc introspection endpoint to the appliance console CLI. Combined with other PRs in other github repos support for the ManageIQ API  will be moved out of code and into the Apache configuration.

**Dependent PRs**
https://github.com/ManageIQ/manageiq-appliance/pull/282
https://github.com/ManageIQ/manageiq-api/pull/828
https://github.com/ManageIQ/manageiq/pull/20131

To Test:

Run the command of the form

```bash
appliance_console_cli \
  --oidc-config \
  --oidc-url="http://my-keycloak.example.com:8080/auth/realms/miq/.well-known/openid-configuration" \
  --oidc-client-host="my-miq.example.com" \
  --oidc-client-id="my-oidc-client"  \
  --oidc-client-secret="********" \
  --oidc-introspection-endpoint="https://my-keycloak.example.com:8443/auth/realms/miq/protocol/openid-connect/token/introspect"
```

Fryguy commented:
All dependent PRs merged.

Fryguy commented on lib/manageiq/appliance_console/oidc_authentication.rb:
```diff
@@ -74,6 +76,7 @@ def validate_oidc_options
         raise "Must specify the OpenID-Connect Provider URL via --oidc-url" if options[:oidc_url].blank?
         raise "Must specify the OpenID-Connect Client ID via --oidc-client-id" if options[:oidc_client_id].blank?
         raise "Must specify the OpenID-Connect Client Secret via --oidc-client-secret" if options[:oidc_client_secret].blank?
+        raise "Must specify the OpenID-Connect Client Introspection Endpoint --oidc-introspection-endpoint" if options[:oidc_introspection_endpoint].blank?
```
Why are we making it required?  I would expect that most providers have a well known endpoint, so I would argue this is optional, and if specified then we include the line.  @abellotti Thoughts?

abellotti replied:
Any objections/issues to having the CLI attempt to fetch the provider metadata at the url specified ?  If we can, then we can simply get the exact introspection URL needed, this would work for all providers and not just keycloak (as well as future IAM where the provider metadata would be specified).

so enhanced logic would be:

- if introspection url not specified:
  - try to fetch the provider metadata, if successful, use the introspection value there.
  - if unsuccessful, try to derive as done in this PR.
  - if cannot derive, fail by requiring the option.

jvlcek replied:
> Any objections/issues to having the CLI attempt to fetch the provider metadata at the url specified ? If we can, then we can simply get the exact introspection URL needed, this would work for all providers and not just keycloak (as well as future IAM where the provider metadata would be specified).
> 
> so enhanced logic would be:
> 
> * if introspection url not specified:
>   
>   * try to fetch the provider metadata, if successful, use the introspection value there.
>   * if unsuccessful, try to derive as done in this PR.
>   * if cannot derive, fail by requiring the option.

Are you suggesting something like this:

```bash
curl -k -L -X GET     -H 'Content-Type: application/x-www-form-urlencoded'     -d 'grant_type=password'     -d 'client_id=${options[:oidc_client_id]}'     -d 'client_secret=${options[:oidc_client_secret]}'  ${options[:oidc_url]}  | sed -e 's/[{}]/''/g'  | sed -s 's/.*introspection_endpoint\":\"//' | sed -s 's/.\".*$//'
```

I do not think we should rely on something like `jq` being there to parse the JSON.

I'm a little concerned that parsing the JSON might be more trouble than the outcome would be worth.

Thoughts?

jvlcek commented:
@abellotti and @Fryguy  oidc_introspection_endpoint is no longer required if the metadata URL can be used to derive it. I think I have addressed all the issues that have been raised.

Thank you again for the input.

Fryguy commented:
@jvlcek @abellotti If a user does _not_ pass the token introspection endpoint and we can derive it, do we still put that value in the conf file?  Is it required in the conf file?

abellotti commented:
I think so, The spec says it is mandatory here https://github.com/zmartzone/mod_auth_openidc/blob/9f1881fb07b8c4fbf8d5a4ce9be6c191f812befb/auth_openidc.conf#L320

@jvlcek have we tried without it  but with the OIDCOAuthClientID and OIDCOAUTHClientSecret specified ?

jvlcek commented:
> I think so, The spec says it is mandatory here https://github.com/zmartzone/mod_auth_openidc/blob/9f1881fb07b8c4fbf8d5a4ce9be6c191f812befb/auth_openidc.conf#L320
> 
> @jvlcek have we tried without it but with the OIDCOAuthClientID and OIDCOAUTHClientSecret specified ?

I tried it just now, just to be sure.

Yes OIDCOAuthIntrospectionEndpoint is mandatory. 
Things fail miserably without   OIDCOAuthIntrospectionEndpoint even if OIDCOAuthClientID and OIDCOAUTHClientSecret are specified.

Fryguy commented:
@bdunne Can you cut a release of this to include this change?

bdunne commented:
> @bdunne Can you cut a release of this to include this change?

Released v5.5.0

chessbyte commented:
@Fryguy @bdunne @jvlcek @abellotti Created https://github.com/ManageIQ/manageiq-appliance/pull/284 to use the newly released gemfile.
